imap/guesstz.c

```diff
@@ -0,0 +1,914 @@
+ *
+ * Copyright (c) 1994-2021 Carnegie Mellon University.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *
+ * 3. The name "Carnegie Mellon University" must not be used to
+ *    endorse or promote products derived from this software without
+ *    prior written permission. For permission or any legal
+ *    details, please contact
+ *      Carnegie Mellon University
+ *      Center for Technology Transfer and Enterprise Creation
+ *      4615 Forbes Avenue
+ *      Suite 302
+ *      Pittsburgh, PA  15213
+ *      (412) 268-7393, fax: (412) 268-7395
+ *      innovation@andrew.cmu.edu
+ *
+ * 4. Redistributions of any form whatsoever must retain the following
+ *    acknowledgment:
+ *    "This product includes software developed by Computing Services
+ *     at Carnegie Mellon University (http://www.cmu.edu/computing/)."
+ *
+ * CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO
+ * THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
+ * AND FITNESS, IN NO EVENT SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE
+ * FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
+ * AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
+ * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include <config.h>
+#include <fts.h>
+#include <syslog.h>
+
+#include <jansson.h>
+
+#include "caldav_db.h"
+#include "dynarray.h"
+#include "libconfig.h"
+#include "ical_support.h"
+
+#include "guesstz.h"
+
+#define GUESSTZDB_BACKEND "twoskip"
+
+static uint8_t guesstzdb_version = 1;
+
+struct obsrec {
+    int64_t onset;
+    int32_t offset;
+};
+
+struct offsetrec {
+    int32_t offset;
+    uint32_t tznum;
+};
+
+struct guesstzdb {
+    struct db *db;
+    struct txn *tid;
+    icaltimetype dbstart;
+    icaltimetype dbend;
+    char *ianaversion;
+    char *fname;
+    char version;
+    time_t created;
+    struct buf buf1;
+    struct buf buf2;
+};
+
+static int format_offset(int32_t offset, char *buffer)
+{
+    char sign = '+';
+    if (offset < 0) {
+        offset = -offset;
+        sign = '-';
+    }
+
+    int hours = offset / 3600;
+    int minutes = (offset % 3600) / 60;
+    int seconds = offset % 60;
+
+    if (hours < 0 || hours > 23 || minutes < 0 || minutes > 60 ||
+            seconds < 0 || seconds > 60) {
+        return 0;
+    }
```

Comment:
Will these ever be < 0?  offset is always > 0 at this point

Reply:
Right, I updated this to only check positive numbers. I also changed it to reject minutes or seconds > 59